File: supporter-product-data/src/main/scala/com/gu/services/AlarmService.scala
package com.gu.services

import com.gu.aws.AwsCloudWatchMetricPut
import com.gu.aws.AwsCloudWatchMetricPut.{
  MetricDimensionName,
  MetricDimensionValue,
  MetricName,
  MetricNamespace,
  MetricRequest,
}
import com.gu.supporterdata.model.Stage
import scala.util.Try

class AlarmService(stage: Stage) {

  def triggerCsvReadAlarm: Try[Unit] = {
    AwsCloudWatchMetricPut(AwsCloudWatchMetricPut.client)(
      MetricRequest(
        MetricNamespace("supporter-product-data"),
        MetricName("CsvReadFailure"),
        Map(MetricDimensionName("Stage") -> MetricDimensionValue(stage.value)),
      ),
    )
  }

  def triggerDynamoWriteAlarm: Try[Unit] = {
    AwsCloudWatchMetricPut(AwsCloudWatchMetricPut.client)(
      MetricRequest(
        MetricNamespace("supporter-product-data"),
        MetricName("DynamoWriteFailure"),
        Map(MetricDimensionName("Stage") -> MetricDimensionValue(stage.value)),
      ),
    )
  }

  def triggerSQSWriteAlarm: Try[Unit] = {
    AwsCloudWatchMetricPut(AwsCloudWatchMetricPut.client)(
      MetricRequest(
        MetricNamespace("supporter-product-data"),
        MetricName("SqsWriteFailure"),
        Map(MetricDimensionName("Stage") -> MetricDimensionValue(stage.value)),
      ),
    )
  }
}

object AlarmService {
  def apply(stage: Stage) = new AlarmService(stage)
}

File: supporter-product-data-typescript/src/typescript/services/alarmService.ts
import {
  CloudWatchClient,
  PutMetricDataCommand,
} from "@aws-sdk/client-cloudwatch";
import { defaultProvider } from "@aws-sdk/credential-provider-node";
import type { Stage } from "../model/stage";

export class AlarmService {
  constructor(
    private readonly stage: Stage,
    private readonly client = new CloudWatchClient({
      region: "eu-west-1",
      credentials: defaultProvider(),
    })
  ) {}

  private async trigger(metricName: string): Promise<void> {
    await this.client.send(
      new PutMetricDataCommand({
        Namespace: "supporter-product-data",
        MetricData: [
          {
            MetricName: metricName,
            Value: 1,
            Unit: "Count",
            Dimensions: [{ Name: "Stage", Value: this.stage }],
          },
        ],
      })
    );
  }

  async triggerCsvReadAlarm(): Promise<void> {
    await this.trigger("CsvReadFailure");
  }

  async triggerDynamoWriteAlarm(): Promise<void> {
    await this.trigger("DynamoWriteFailure");
  }

  async triggerSqsWriteAlarm(): Promise<void> {
    await this.trigger("SqsWriteFailure");
  }
}
